src/app/InteractionModelEngine.cpp

```diff
@@ -240,9 +240,78 @@ DispatchSingleClusterCommand(chip::ClusterId aClusterId, chip::CommandId aComman
         "Default DispatchSingleClusterCommand is called, this should be replaced by actual dispatched for cluster commands");
 }
 
+CHIP_ERROR __attribute__((weak))
+ReadSingleClusterData(NodeId aNodeId, ClusterId aClusterId, EndpointId aEndPointId, FieldId aFieldId, TLV::TLVWriter & aWriter)
+{
+    ChipLogDetail(DataManagement,
+                  "Received Cluster Command: Cluster=%" PRIx16 " NodeId=%" PRIx64 " Endpoint=%" PRIx8 " FieldId=%" PRIx8,
+                  aClusterId, aNodeId, aEndPointId, aFieldId);
+    ChipLogError(DataManagement,
+                 "Default ReadSingleClusterData is called, this should be replaced by actual dispatched for cluster");
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR __attribute__((weak))
+WriteSingleClusterData(NodeId aNodeId, ClusterId aClusterId, EndpointId aEndPointId, FieldId aFieldId, TLV::TLVReader & aReader)
+{
+    ChipLogDetail(DataManagement,
+                  "Received Cluster Command: Cluster=%" PRIx16 " NodeId=%" PRIx64 " Endpoint=%" PRIx8 " FieldId=%" PRIx8,
+                  aClusterId, aNodeId, aEndPointId, aFieldId);
+    ChipLogError(DataManagement,
+                 "Default WriteSingleClusterData is called, this should be replaced by actual dispatched for cluster");
+    return CHIP_NO_ERROR;
+}
+
 uint16_t InteractionModelEngine::GetReadClientArrayIndex(const ReadClient * const apReadClient) const
 {
     return static_cast<uint16_t>(apReadClient - mReadClients);
 }
+
+CHIP_ERROR InteractionModelEngine::GetFirstAvailableClusterInfo(ClusterInfo *& apClusterInfo)
+{
+    if (mNumClusterInfos >= IM_SERVER_MAX_NUM_PATH_GROUPS)
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+    apClusterInfo = &mClusterInfoPool[mNumClusterInfos];
+    mNumClusterInfos++;
+    apClusterInfo->ClearDirty();
+    return CHIP_NO_ERROR;
+}
+
+void InteractionModelEngine::ReleaseClusterInfoList(ClusterInfo * apClusterInfo)
+{
+    ClusterInfo * clusterInfo = apClusterInfo;
+    while (clusterInfo != nullptr)
+    {
+        ClusterInfo * current = clusterInfo;
+        clusterInfo           = current->mpNext;
+        current->mDirty       = false;
+        current->mpNext       = nullptr;
+    }
+}
+
+CHIP_ERROR InteractionModelEngine::AcquireClusterInfo(ClusterInfo *& aClusterInfo, AttributePathParams & aAttributePathParams)
+{
+    for (uint32_t index = 0; index < IM_SERVER_MAX_NUM_PATH_GROUPS; index++)
```

Comment:
How about keeping a separate "Free list start" and initializing that in the constructor?

That seems to be much faster to aquire (no more loop to MAX_NUM_PATH_GROUPS) and would not rely on the 'is dirty' flag anymore (which I am unsure if it is correct ... I see set/clear dirty in other code paths as well)

Reply:
Agree, Done, thanks